compiler/rmi/rmi-interface/src/org/jetbrains/kotlin/rmi/DaemonParams.kt

```diff
@@ -0,0 +1,183 @@
+
+fun Iterable<String>.propParseFilter(parsers: List<PropParser<*,*,*>>) : Iterable<String>  {
+    var currentParser: PropParser<*,*,*>? = null
+    return filter { param ->
+        if (currentParser == null) {
+            currentParser = parsers.find { param.equals("--daemon-" + it.prop.name) }
+            if (currentParser != null) {
+                if (currentParser is BoolPropParser<*,*>) {
+                    currentParser!!.apply("")
+                    currentParser = null
+                }
+                false
+            }
+            else true
+        }
+        else {
+            currentParser!!.apply(param)
+            currentParser = null
+            false
+        }
+    }
+}
+
+// TODO: find out how to create more generic variant using first constructor
+//fun<C> C.propsToParams() {
+//    val kc = C::class
+//    kc.constructors.first().
+//}
+
+public interface CmdlineParams : Serializable {
+    public val asParams: Iterable<String>
+    public val parsers: List<PropParser<*,*,*>>
+}
+
+public fun Iterable<String>.propParseFilter(vararg cs: CmdlineParams) : Iterable<String> =
+    propParseFilter(cs.flatMap { it.parsers })
+
+
+public data class DaemonLaunchingOptions(
+        public var jvmParams: List<String> = listOf()
+) : CmdlineParams {
+
+    override val asParams: Iterable<String>
+        get() =
+            propToParams(::jvmParams, { it.joinToString("##") }) // TODO: consider some other options rather than using potentially dangerous delimiter
+
+    override val parsers: List<PropParser<*,*,*>>
+        get() = listOf( PropParser(this, ::jvmParams, { it.split("##")})) // TODO: see appropriate comment in asParams
+}
+
+public data class DaemonOptions(
+        public var port: Int = COMPILE_DAEMON_DEFAULT_PORT,
+        public var autoshutdownMemoryThreshold: Long = 0 /* 0 means unchecked */,
+        public var autoshutdownIdleSeconds: Int = 0 /* 0 means unchecked */,
+        public var startEcho: String = COMPILER_SERVICE_RMI_NAME
+) : CmdlineParams {
+
+    override val asParams: Iterable<String>
+        get() =
+            propToParams(::port) +
+            propToParams(::autoshutdownMemoryThreshold) +
+            propToParams(::autoshutdownIdleSeconds) +
+            propToParams(::startEcho)
+
+    override val parsers: List<PropParser<*,*,*>>
+            get() = listOf( PropParser(this, ::port, { it.toInt()}),
+                            PropParser(this, ::autoshutdownMemoryThreshold, { it.toLong()}),
+                            PropParser(this, ::autoshutdownIdleSeconds, { it.toInt()}),
+                            PropParser(this, ::startEcho, { it.trim('"') }))
+}
+
+
+val COMPILER_ID_DIGEST = "MD5"
+
+
+fun updateSingleFileDigest(file: File, md: MessageDigest) {
+    val stream = DigestInputStream(file.inputStream(), md)
+    val buf = ByteArray(1024)
+    while (stream.read(buf) == buf.size()) {}
+    stream.close()
```

Comment:
shouldn't we use [`use` function](http://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/use.html) here:

```
DigestInputStream(file.inputStream(), md).use { stream ->
    // do stuff
}
```